test/evaluate/folding03.f90

```diff
@@ -0,0 +1,305 @@
+! Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+!
+! Licensed under the Apache License, Version 2.0 (the "License");
+! you may not use this file except in compliance with the License.
+! You may obtain a copy of the License at
+!
+!     http://www.apache.org/licenses/LICENSE-2.0
+!
+! Unless required by applicable law or agreed to in writing, software
+! distributed under the License is distributed on an "AS IS" BASIS,
+! WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+! See the License for the specific language governing permissions and
+! limitations under the License.
+
+! Test operation folding edge case (both expected value and messages)
+! These tests make assumptions regarding real(4) and integer(4) extrema.
+
+#define TEST_ISNAN(v) logical, parameter :: test_##v =.NOT.(v.EQ.v)
+
+! TODO: 2 behaviors to clarify:
```

Comment:
Testing of edge-cases raised two questions detailed as a TODO in folding03.f90 test cases: I think integer divide by zero should be made a fatal error (I have not made the change so far). Also, is -2147483648_4 literal not supported on purpose ?

Reply:
Unfortunately, Fortran doesn't have signed integer literals in expressions, so `-2147483648_4` is actually a negation operation `-(2147483648_4)`.

Division by zero during folding should (I think) be a warning, not a fatal error, since it can be possible that the value of the expression is not needed at runtime.